File: templates/knapsack.py
def ss(target, B):
    import array
    possible = array.array('b', [1] + [0]*target)

    for b in B:
        next_possible = array.array('b', possible)
        for i in range(target):
            if possible[i] == 1 and i + b <= target:
                next_possible[i+b] = 1
        possible = next_possible

    return possible

def ss_ncp(target, B):
    import array
    possible = array.array('b', [1] + [0]*target)

    for b in B:
        for mass in range(target, -1, -1):
            if possible[mass] == 1 and mass + b <= target:
                possible[mass+b] = 1
    return possible

File: templates/knapsack.cpp
// tested on https://atcoder.jp/contests/abc056/tasks/arc070_b
vector<bool> ss_ncp(int target, vector<int> B) {
  auto possible = vector<bool>(target+1, false);
  possible[0] = true;

  for (auto b : B) {
    for (int mass=target; mass!=-1; mass--) {
      if (possible[mass] && (mass + b <= target))
          possible[mass+b] = true;
      }
  } 

  return possible;
}
